openscm/core.py

```diff
@@ -1,24 +1,304 @@
 """
-Then OpenSCM low-level API includes the basic functionality to run a
-particular simple climate model with OpenSCM as well as
-setting/getting its parameter values. Mapping of parameter names and
-units is done internally.
+The OpenSCM low-level API includes the basic functionality to run a particular
+simple climate model with OpenSCM as well as setting/getting its parameter values.
+Mapping of parameter names and units is done internally.
 
-Parts of this API definition seems unpythonic as it is designed to
-be easily implementable in several programming languages.
+Parts of this API definition seems unpythonic as it is designed to be easily
+implementable in several programming languages.
 """
 
-from typing import Sequence, Tuple
+from enum import Enum
+from typing import Any, Dict, Sequence, Tuple
 
 
-class ParameterLengthError(Exception):
+class ParameterType(Enum):
+    """
+    Parameter type.
+    """
+
+    SCALAR = 1
+    TIMESERIES = 2
+
+
+class _Parameter:
+    """
+    Represents a parameter in the parameter hierarchy.
+    """
+
+    _children: Dict[str, "_Parameter"]
+    """Child parameters"""
+
+    _data: Any
+    """Data"""
+
+    _has_been_aggregated: bool
+    """
+    Tells if parameter has already been read in an aggregated way, i.e., aggregating
+    over child parameters
+    """
+
+    _has_been_written_to: bool
+    """Tells if parameter data has already been changed"""
+
+    _name: str
+    """Name"""
+
+    _parent: "_Parameter"
+    """Parent parameter"""
+
+    _type: ParameterType
+    """Parameter type"""
+
+    _unit: str
+    """Unit"""
+
+    def __init__(self, name: str):
+        """
+        Initialize.
+
+        Parameters
+        ----------
+        name
+            Name
+        """
+        self._children = {}
+        self._has_been_aggregated = False
+        self._has_been_written_to = False
+        self._name = name
+        self._parent = None
+        self._type = None
+        self._unit = None
+
+    def get_or_create_child_parameter(
+        self, name: str, unit: str, parameter_type: ParameterType
+    ) -> "_Parameter":
+        """
+        Get a (direct) child parameter of this parameter. Create and add it if not
+        found.
+
+        Parameters
+        ----------
+        name
+            Name
+        unit
+            Unit
+        parameter_type
+            Parameter type
+
+        Raises
+        ------
+        RegionAggregatedError
+            If the subregion would need to be added and a parameter of this region has
```

Comment:
```suggestion
            If the child would need to be added and this parameter has
```

Reply:
Hmm, I would prefer to call "child regions" "subregions", but I am not strict about this. But it should read "a parameter of this region", because a region does not represent a parameter, but collects all of those for the specific region...

Reply:
I also missed the context here, sorry. done